File: app/src/main/java/hottopic/mit/co/nz/cleaningservice/network/ServiceStatusChangedRequest.java
package hottopic.mit.co.nz.cleaningservice.network;

import android.content.Context;

import java.util.Map;

import hottopic.mit.co.nz.cleaningservice.Constants;
import hottopic.mit.co.nz.cleaningservice.entities.network.response.BooleanResponse;
import hottopic.mit.co.nz.cleaningservice.entities.orders.Order;
import hottopic.mit.co.nz.cleaningservice.network.service.OrderService;
import io.reactivex.Single;

public class ServiceStatusChangedRequest extends BaseRequest<OrderService, BooleanResponse, BooleanResponse> {
    private int type;
    private Order order;

    public ServiceStatusChangedRequest(Context context, int type, Order order) {
        super(context);
        this.type = type;
        this.order = order;
    }

    @Override
    protected Map<String, String> getHeader() {
        return null;
    }

    @Override
    protected Class<OrderService> getEndpointClass() {
        return OrderService.class;
    }

    @Override
    protected Single<BooleanResponse> getEndpoint(OrderService endpoint) {
        if (Constants.TYPE_SERVICE_STARTED == type) {
            return endpoint.orderStarted(order);
        }else{
            return endpoint.orderfinished(order);
        }
    }

    @Override
    protected BooleanResponse dealWithResponse(BooleanResponse response) {
        return response;
    }
}

File: app/src/main/java/hottopic/mit/co/nz/cleaningservice/network/ServiceStatusChangedRequest.kt
package hottopic.mit.co.nz.cleaningservice.network

import android.content.Context

import hottopic.mit.co.nz.cleaningservice.Constants
import hottopic.mit.co.nz.cleaningservice.entities.network.response.BooleanResponse
import hottopic.mit.co.nz.cleaningservice.entities.orders.Order
import hottopic.mit.co.nz.cleaningservice.network.service.OrderService
import io.reactivex.Single

class ServiceStatusChangedRequest(context: Context, private val type: Int, private val order: Order) : BaseRequest<OrderService, BooleanResponse, BooleanResponse>(context) {

    override val header: Map<String, String>?
        get() = null

    override val endpointClass: Class<OrderService>
        get() = OrderService::class.java

    override fun getEndpoint(endpoint: OrderService?): Single<BooleanResponse?>? {
        return if (Constants.TYPE_SERVICE_STARTED == type) {
            endpoint?.orderStarted(order)
        } else {
            endpoint?.orderfinished(order)
        }
    }

    override fun dealWithResponse(response: BooleanResponse): BooleanResponse? {
        return response
    }
}
